add support for custom merge strategy

oldgromit opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Adding custom merge strategy to himl so default behavior can be customized. 
<!--- Describe your changes in detail -->
Currently ConfigProcessor builds a Merger using 

`merger = Merger([(list, ["append"]), (dict, ["merge"])], ["override"], ["override"])`

We want to add support for custom merge strategies.  i.e. for a list of items, we want to be able to override the default values, instead of default append behavior. Also we would like to be able to remove certain default values from lower nodes in the hierarchy.

We can change ConfigProcessor and ConfigGenerator init method signature so those optional merge strategies can be passed in, and send to deepmerge Merger object.

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
We want to define certain default values for overall hierarchy, but also allows flexibility to override or remove default values by lower nodes if necessary. Currently the merging policy/strategy is hardcoded in ConfigGenerator. We'd like to propose to change that as an optional parameter so user can pass in customized functions if they have that need.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
1. Modified main.py so an example custom merge strategy as a function is defined, and this function is passed in as an optional parameter to ConfigProcessor
2. Modify default.yaml and cluster.yaml to provide an example on how to use the custom merge strategy. User can define different unique keys for each record to distinguish records. Also a "remove" flag is provided in the example to showcase how to delete a default value.
3. Run `himl examples/complex/env=dev/region=us-east-1/cluster=cluster2 ` to test expected behavior: 
 * records with same 'id' field are overridden 
 * record with 'remove' flag is removed
4. Run the same command against other folder such as `himl examples/complex/env=dev/region=us-east-1/cluster=cluster1` to make sure default behavior is not changed if no 'id' field is defined.


## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [X] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

amuraru commented on himl/main.py:
```diff
@@ -31,7 +47,8 @@ def do_run(self, opts):
                                  opts.output_format, opts.print_data, opts.output_file, opts.skip_interpolation_resolving,
-                                 opts.skip_interpolation_validation, opts.skip_secrets, opts.multi_line_string)
+                                 opts.skip_interpolation_validation, opts.skip_secrets, opts.multi_line_string,
+                                 type_strategies= [(list, [self.strategy_merge_override,"append"]), (dict, ["merge"])] )
```
This is actually changing the current behaviour by setting the new strategy as default.
Please add a new argument "use-merge-list-strategy" and add this if that is set else keep the old behaviour

amuraru replied:
@oldgromit what I meant is that `himl/main.py` is not really just an example but the main entry point in the `himl` binary installed by the python package:
https://github.com/adobe/himl/blob/master/setup.py#L59

So to avoid breaking existing clients using this with the default strategy I suggested you add a new argument here: https://github.com/adobe/himl/blob/master/himl/main.py#L66-L68
`--use-merge-list-strategy` and when only that is set configure the new strategy.

oldgromit replied:
I see what you mean. my understanding of main.py as an example is misguided. 
I can definitely add a new command line argument for the strategy. However, the new strategy would be a function that should be implemented and provided by the user, not himl. is that ok we dynamically load the user specified module and pass the function to Merger, if the optional argument is passed?

costimuraru replied:
@oldgromit I'm wondering what would be the best way for you override the merging strategy. 

1. Are you using `himl` as a standalone binary? In that case, it might be useful for you to have an argument like @amuraru mentioned, where you can control the merging behavior. The way I see it, the argument could be a `str` and receive one of the 3 values:
`--merge-list-strategy=append/override/prepend` (see [docs](https://deepmerge.readthedocs.io/_/downloads/en/latest/pdf/)), where _append_ is the default:

```python
parser.add_argument('--merge-list-strategy',
                    choices=['append', 'override', 'prepend'],
                    help='...')
```

Then you can run himl like this to achieve your goal:
```python
himl --merge-list-strategy=override ...
```

For this to work, you can [add the argument](https://github.com/adobe/himl/blob/master/himl/main.py#L66-L68) and then update this line of code to something like:
```
type_strategies= [(list, [opts.merge_list_strategy]), (dict, ["merge"])] )
```

In this way, we pass the value received from the user (override/append/prepend) to the deepmerge lib.

PS: Not sure if it's worth adding an argument for the dict (`--merge-dict-strategy=merge/override`) as well, probably not atm.

2. If you are using `himl` as a python module, it looks like you can pass in the desired strategies with the code you've written so far, correct?

oldgromit replied:
thanks @costimuraru  for your feedback. you brought up an interesting point I haven't considered. the 3 values (append/override/prepend) are 3 merge strategies deepmerge provides out of box, we should allow users to specify which one they want to use. I've updated the code to offer that flexibility. 

The original goal of my change is to allow a user specified merge strategy that can be passed to himl. a custom function that user has total control of. the example I provided merges/overrides depends on an "id" attribute, user can specify any other combo of fields to uniquely identifying records, they have total control of merging behavior.

I've pushed a new PR to illustrate how user can build their own custom function and himl dynamically loads specified module and function and pass it as the new merge strategy. Currently it only works with list data type.

Regarding #2 you mentioned, yes, if user have access to ConfigProcessor directly, they have a lot of more freedom and can pass in more elaborate merge strategies. 

@amuraru @costimuraru  Please review the latest PR and provide your valuable feedback. I'll update the documentation once PR is approved. Thanks!

amuraru commented on himl/main.py:
```diff
@@ -21,6 +21,22 @@ def run(self, args):
+        result = deepcopy(base)
+        for nxto in nxt:
+            for baseo in result:
+                if 'id' in baseo and 'id' in nxto and baseo['id'] == nxto['id']:
```
What is the list items are nort dicts? But say plain string. Should we add a check for that?

oldgromit replied:
done. added type check to avoid this edge case:
 ```
           # if list is not a list of dicts, bail out and let the next strategy to execute
            if not isinstance(baseo,dict) or not isinstance(nxto,dict):
                return STRATEGY_END
```

amuraru reviewed (commented):
Thanks for your PR - see my comments

amuraru commented on himl/main.py:
```diff
@@ -21,6 +21,31 @@ def run(self, args):
+                # if list is a list of strings, simply compare string values, and don't duplicate strings
+                # this behavior is actually the same as the default append_unique (using set) strategy 
+                # but unfortuntely deepmerge does not provide chained executed strategies, only fallback strategies
+                if isinstance(baseo,str) and isinstance(nxto,str) and baseo == nxto:
```
this should be on line 35 - otherwise a list like:
```
myList:
 - id1
 - id2
 - id3
```

would crash in flames ;)

oldgromit replied:
Good catch. Will update. Should we leave this custom function in himl package? maybe as an example to showcase how to build a custom merging strategy? Will check in to tie everything together.

amuraru commented:
@danielcoman would you please review this as well?

costimuraru reviewed (commented):
Let's also update the README.md with some docs on this new capability you've added (what it does, how it can be used), @oldgromit.

oldgromit commented on examples/complex/env=dev/region=us-east-1/cluster=cluster2/cluster.yaml:
```diff
@@ -1 +1,16 @@
+    value: 5
+metrics:
+  - cpu
+  - exec
```
missed this one. will update for next PR.

oldgromit replied:
done in the latest PR. please review. @costimuraru

amuraru commented on himl/main.py:
```diff
@@ -63,6 +75,8 @@ def get_parser(parser=None):
                             help='the working directory')
         parser.add_argument('--multi-line-string', action='store_true',
                             help='will overwrite the global yaml dumper to use block style')
```
Use an enum choice here please: 
https://stackoverflow.com/a/46385352/649647

oldgromit replied:
Thanks. It makes sense to put allowable values (append/override/prepend/append_unique) as enum. Just a clarification.. As the new argument --merge-list-strategy takes two forms of format, 

1. --merge-list-strategy default append/override/prepend/append_unique (predefined merge strategies deepmerge provides)
2. --merge-list-strategy user_defined_module user_defined_function (user defined)

The example you gave will not cover both scenarios. Do you suggest to just define an enum for those predefined strategies and just check them during parsing, or should we split merge-list-strategy argument to two separate arguments, one to take deepmerge defined strategies, one to custom strategies? for examp,e --merge-list-default-strategy  and merge-list-custom-strategy ?

Thanks.

oldgromit replied:
@amuraru  pushed a PR for implementation #1, defined an enum for 4 default merge strategies and validate passed in value, and throw a meaningful error msg. 
plz let me know if #2 implementation makes more sense and it's a better alternative.

amuraru replied:
see my comment above - 2# doesn't make sense to me here

amuraru commented on himl/config_generator.py:
```diff
@@ -28,9 +28,11 @@
+                type_strategies = [(list, ["append"]), (dict, ["merge"])], fallback_strategies = ["override"], type_conflict_strategies = ["override"]):
 
         path = self.get_relative_path(path)
+       
```
nit: remove

oldgromit replied:
done

amuraru commented on himl/main.py:
```diff
@@ -28,10 +46,25 @@ def do_run(self, opts):
+                    print(err)
+                    return       
+            else:
+                imported_module = importlib.import_module(opts.merge_list_strategy[0])
```
I don't think importing dynamically an external module is practical for users that are consuming himl as a cli.
Think a user using https://github.com/adobe/himl#using-docker-image

We should remove this option (of providing an external module as a paramter) from this cli and instead allow this only for users consuming himl as a library

oldgromit replied:
I agree. Initially I just used main.py as an example to showcase how to use it. If we use main.py as strictly as a cli, it does not make sense. I have removed the option to provide an external module. We will defer documentation for using custom module to README file.

oldgromit replied:
@amuraru  @costimuraru Pushed latest code.  Removed dynamic loading of external module, and updated README to describe how to pass custom functions to ConfigProcessor.  Please review. Thanks.

amuraru commented on himl/main.py:
```diff
@@ -63,6 +96,8 @@ def get_parser(parser=None):
                             help='the working directory')
         parser.add_argument('--multi-line-string', action='store_true',
                             help='will overwrite the global yaml dumper to use block style')
+        parser.add_argument('--merge-list-strategy', dest='merge_list_strategy', nargs=2,
```
let's define as `type= DefaultMergeStrategy, choices=list(DefaultMergeStrategy)`

oldgromit replied:
Done.

oldgromit replied:
also renamed the argument name to list-merge-strategy for consistency. 
Pushed latest change. Plz review.

amuraru commented on himl/main.py:
```diff
@@ -10,9 +10,19 @@
 import argparse
 import os
-
+from copy import deepcopy
```
Not needed

amuraru commented on himl/main.py:
```diff
@@ -10,9 +10,19 @@
-
+from copy import deepcopy
 from .config_generator import ConfigProcessor
+import importlib
```
Same - not needed

amuraru commented on himl/main.py:
```diff
@@ -10,9 +10,19 @@
+import importlib
+from enum import Enum
+
+class DefaultMergeStrategy(Enum):
```
```suggestion
class ListMergeStrategy(Enum):
```

costimuraru commented on himl/custom_merger_example.py:
```diff
@@ -0,0 +1,29 @@
+                                 opts.skip_interpolation_validation, opts.skip_secrets, opts.multi_line_string,
+                                 type_strategies= [(list, [strategy_merge_override,'append']), (dict, ["merge"])] )
+"""
+def strategy_merge_override(config, path, base, nxt):
```
I see this is described very well in the [README.md](https://github.com/adobe/himl/pull/104/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R365). Is there a reason for keeping here as well or we can remoce it?
I'm thinking this file will be bundled in the python package and my concern is that we are polluting it with code that's not actually used. What do you think?

oldgromit replied:
kept it just in case people don't read README :) 
I do agree we shouldn't keep it there to keep bundle clean and mean. Removed.

There's a test_config_generator.py in tests folder. As far as I see, it's not used.  Should I remove that as well?

costimuraru commented on README.md:
```diff
@@ -352,3 +353,38 @@ VA7:     !include configs/env=int/region=va7/kafka-brokers.yaml regionBrokers.VA
 
 This will replace the value after interpolation with the value of the regionBrokers.VA7 found under the configs/env=int/region=va7/kafka-brokers.yaml path.
+
+## custom merge strategy
```
Minor: To be consistent with the other sections, let's start the section name with a capital letter: `Custom merge strategy`

oldgromit replied:
Done.

costimuraru commented:
@oldgromit this is looking quite nice. Added just a couple more comments. Let me know what you think.

oldgromit commented:
> @oldgromit this is looking quite nice. Added just a couple more comments. Let me know what you think.

thanks.  I agree it's important to keep core lib clean, so I removed the example code, and simply will rely on the README for people to figure out the new feature.

oldgromit commented:
@amuraru @costimuraru Thanks for all the great feedback. I've throughly enjoyed our discussions and have incorporated those feedback.  This has been an interesting exchange of ideas. Would appreciate it if you guys can merge the new feature into master branch and  release it so my team can leverage the new feature. Thanks again for your great work on this very useful package!
